## Import Libraries

In [1]:
import pandas as pd
import plotly.graph_objects as go

#Set float number just have two decimals
pd.set_option('display.float_format','{:.2f}' .format)
#Set option display max columns
pd.options.display.max_columns=999

## Load and Concat Dataset

In [2]:
df1 = pd.read_csv('Premier_league_18_19.csv')
df2 = pd.read_csv('Premier_league_19_20.csv')
df3 = pd.read_csv('Premier_league_20_21.csv')

In [3]:
df = pd.concat([df1,df2,df3])
df.head()

date          home_team        away_team  home_score  away_score  \
0  08/12/2018            Arsenal  Manchester City           0           2   
1  08/11/2018    AFC Bournemouth     Cardiff City           2           0   
2  08/11/2018             Fulham   Crystal Palace           0           2   
3  08/11/2018  Huddersfield Town          Chelsea           0           3   
4  08/12/2018          Liverpool  West Ham United           4           0   

   home_possession_%  away_possession_%  home_shots_on_target  \
0              42.00              58.00                  3.00   
1              62.90              37.10                  4.00   
2              66.30              33.70                  6.00   
3              37.20              62.80                  1.00   
4              64.80              35.20                  8.00   

   away_shots_on_target  home_shots  away_shots  home_touches  away_touches  \
0                  8.00        9.00       17.00        593.00        734.00   
1                  1.00       12.00       10.00        711.00        483.00   
2                 10.00       15.00       12.00        893.00        564.00   
3                  4.00        6.00       13.00        550.00        856.00   
4                  2.00       18.00        5.00        843.00        551.00   

   home_passes  away_passes  home_tackles  away_tackles  home_clearances  \
0       401.00       555.00         22.00          6.00            23.00   
1       502.00       287.00         11.00         22.00            37.00   
2       672.00       347.00         25.00         21.00            22.00   
3       372.00       658.00         25.00         12.00            13.00   
4       673.00       360.00         19.00         23.00             8.00   

   away_clearances  home_corners  away_corners  home_offsides  away_offsides  \
0             9.00          2.00          9.00           7.00           2.00   
1            30.00          7.00          4.00           0.00           2.00   
2            30.00          5.00          5.00           2.00           3.00   
3            21.00          2.00          5.00           2.00           1.00   
4            20.00          5.00          4.00           5.00           3.00   

   home_yellow_cards  away_yellow_cards  home_red_cards  away_red_cards  \
0               2.00               2.00            0.00            0.00   
1               1.00               1.00            0.00            0.00   
2               1.00               2.00            0.00            0.00   
3               2.00               1.00            0.00            0.00   
4               1.00               2.00            0.00            0.00   

   home_fouls_conceded  away_fouls_conceded  
0                11.00                14.00  
1                11.00                 9.00  
2                 9.00                11.00  
3                 9.00                 8.00  
4                14.00                 9.00

 ## Data Manipulation

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1140 entries, 0 to 379
Data columns (total 29 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   date                  1140 non-null   object 
 1   home_team             1140 non-null   object 
 2   away_team             1140 non-null   object 
 3   home_score            1140 non-null   int64  
 4   away_score            1140 non-null   int64  
 5   home_possession_%     1140 non-null   float64
 6   away_possession_%     1140 non-null   float64
 7   home_shots_on_target  1140 non-null   float64
 8   away_shots_on_target  1140 non-null   float64
 9   home_shots            1140 non-null   float64
 10  away_shots            1140 non-null   float64
 11  home_touches          1140 non-null   float64
 12  away_touches          1140 non-null   float64
 13  home_passes           1140 non-null   float64
 14  away_passes           1140 non-null   float64
 15  home_tackles          

### Handling Datetime Column

In [5]:
#convert 'date' column datatype to datetime
df['date'] = pd.to_datetime(df['date'], format='%m/%d/%Y')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1140 entries, 0 to 379
Data columns (total 29 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   date                  1140 non-null   datetime64[ns]
 1   home_team             1140 non-null   object        
 2   away_team             1140 non-null   object        
 3   home_score            1140 non-null   int64         
 4   away_score            1140 non-null   int64         
 5   home_possession_%     1140 non-null   float64       
 6   away_possession_%     1140 non-null   float64       
 7   home_shots_on_target  1140 non-null   float64       
 8   away_shots_on_target  1140 non-null   float64       
 9   home_shots            1140 non-null   float64       
 10  away_shots            1140 non-null   float64       
 11  home_touches          1140 non-null   float64       
 12  away_touches          1140 non-null   float64       
 13  home_passes        

### Sorting Dataframe 

In [6]:
#sorting dataframe by the date
df = df.sort_values(by='date', ascending=True, ignore_index=True)
df

date                home_team                 away_team  \
0    2018-08-11          AFC Bournemouth              Cardiff City   
1    2018-08-11                   Fulham            Crystal Palace   
2    2018-08-11        Huddersfield Town                   Chelsea   
3    2018-08-11        Manchester United            Leicester City   
4    2018-08-11         Newcastle United         Tottenham Hotspur   
...         ...                      ...                       ...   
1135 2021-05-23                   Fulham          Newcastle United   
1136 2021-05-23                  Arsenal  Brighton and Hove Albion   
1137 2021-05-23          West Ham United               Southampton   
1138 2021-05-23             Leeds United      West Bromwich Albion   
1139 2021-05-23  Wolverhampton Wanderers         Manchester United   

      home_score  away_score  home_possession_%  away_possession_%  \
0              2           0              62.90              37.10   
1              0           2              66.30              33.70   
2              0           3              37.20              62.80   
3              2           1              46.30              53.70   
4              1           2              40.40              59.60   
...          ...         ...                ...                ...   
1135           0           2              64.30              35.70   
1136           2           0              53.90              46.10   
1137           3           0              36.90              63.10   
1138           3           1              64.40              35.60   
1139           1           2              57.20              42.80   

      home_shots_on_target  away_shots_on_target  home_shots  away_shots  \
0                     4.00                  1.00       12.00       10.00   
1                     6.00                 10.00       15.00       12.00   
2                     1.00                  4.00        6.00       13.00   
3                     6.00                  4.00        8.00       13.00   
4                     2.00                  5.00       15.00       15.00   
...                    ...                   ...         ...         ...   
1135                  0.00                  4.00       14.00       10.00   
1136                  5.00                  1.00       16.00        5.00   
1137                  7.00                  5.00       14.00       17.00   
1138                  9.00                  5.00       17.00       14.00   
1139                  4.00                  4.00       14.00        9.00   

      home_touches  away_touches  home_passes  away_passes  home_tackles  \
0           711.00        483.00       502.00       287.00         11.00   
1           893.00        564.00       672.00       347.00         25.00   
2           550.00        856.00       372.00       658.00         25.00   
3           697.00        744.00       485.00       543.00         26.00   
4           622.00        809.00       387.00       573.00         15.00   
...            ...           ...          ...          ...           ...   
1135        778.00        517.00       596.00       333.00         18.00   
1136        727.00        645.00       565.00       489.00         14.00   
1137        483.00        715.00       300.00       534.00         12.00   
1138        685.00        465.00       499.00       268.00         21.00   
1139        744.00        634.00       571.00       440.00         19.00   

      away_tackles  home_clearances  away_clearances  home_corners  \
0            22.00            37.00            30.00          7.00   
1            21.00            22.00            30.00          5.00   
2            12.00            13.00            21.00          2.00   
3            14.00            37.00            14.00          2.00   
4            27.00            17.00            43.00          3.00   
...            ...              ...              ...           ...   
1135         

### Adding Variables

Adding some variables of team performance metrics to be analyzed:
- Shot on Target % (SoT_%) = Shots on Target / Shots (Home/Away)
- Quantity Conversion Rate (QuanCR) = Goals / Shots (Home/Away)
- Quality Conversion Rate (QualCR) = Goals / Shots on Target (Home/Away)
- Points per game (ppg) = 3 if a team win, 1 if a match draw, 0 if a team lose
- Win rate (win_rate) = 100 if a team win in a match, 0 if they lose or draw (Home/Away)

In [7]:
df['home_SoT_%']= df['home_shots_on_target']/df['home_shots']*100
df['away_SoT_%']= df['away_shots_on_target']/df['away_shots']*100
df['home_QuanCR'] = df['home_score']/df['home_shots']*100
df['away_QuanCR'] = df['away_score']/df['away_shots']*100
df['home_QualCR'] = df['home_score']/df['home_shots_on_target']*100
df['away_QualCR'] = df['away_score']/df['away_shots_on_target']*100

for i in range(len(df)):
    if df.loc[i,'home_score']>df.loc[i,'away_score']:
        df.loc[i,'home_ppg'] = 3
        df.loc[i,'away_ppg'] = 0
    elif df.loc[i,'home_score']==df.loc[i,'away_score']:
        df.loc[i,'home_ppg'] = 1
        df.loc[i,'away_ppg'] = 1
    else:
        df.loc[i,'home_ppg'] = 0
        df.loc[i,'away_ppg'] = 3
        
df.loc[df['home_score']>df['away_score'], 'home_win_rate'] = 100
df.loc[df['home_score']<df['away_score'], 'home_win_rate'] = 0
df.loc[df['home_score']==df['away_score'], 'home_win_rate'] = 0

df.loc[df['away_score']>df['home_score'], 'away_win_rate'] = 100
df.loc[df['away_score']<df['home_score'], 'away_win_rate'] = 0
df.loc[df['away_score']==df['home_score'], 'away_win_rate'] = 0
#There will be nan value, a product of 0/0. Replace or fill the na value with 0.
df = df.fillna(0, axis=1)

In [8]:
df.head()

date          home_team          away_team  home_score  away_score  \
0 2018-08-11    AFC Bournemouth       Cardiff City           2           0   
1 2018-08-11             Fulham     Crystal Palace           0           2   
2 2018-08-11  Huddersfield Town            Chelsea           0           3   
3 2018-08-11  Manchester United     Leicester City           2           1   
4 2018-08-11   Newcastle United  Tottenham Hotspur           1           2   

   home_possession_%  away_possession_%  home_shots_on_target  \
0              62.90              37.10                  4.00   
1              66.30              33.70                  6.00   
2              37.20              62.80                  1.00   
3              46.30              53.70                  6.00   
4              40.40              59.60                  2.00   

   away_shots_on_target  home_shots  away_shots  home_touches  away_touches  \
0                  1.00       12.00       10.00        711.00        483.00   
1                 10.00       15.00       12.00        893.00        564.00   
2                  4.00        6.00       13.00        550.00        856.00   
3                  4.00        8.00       13.00        697.00        744.00   
4                  5.00       15.00       15.00        622.00        809.00   

   home_passes  away_passes  home_tackles  away_tackles  home_clearances  \
0       502.00       287.00         11.00         22.00            37.00   
1       672.00       347.00         25.00         21.00            22.00   
2       372.00       658.00         25.00         12.00            13.00   
3       485.00       543.00         26.00         14.00            37.00   
4       387.00       573.00         15.00         27.00            17.00   

   away_clearances  home_corners  away_corners  home_offsides  away_offsides  \
0            30.00          7.00          4.00           0.00           2.00   
1            30.00          5.00          5.00           2.00           3.00   
2            21.00          2.00          5.00           2.00           1.00   
3            14.00          2.00          5.00           4.00           2.00   
4            43.00          3.00          5.00           1.00           0.00   

   home_yellow_cards  away_yellow_cards  home_red_cards  away_red_cards  \
0               1.00               1.00            0.00            0.00   
1               1.00               2.00            0.00            0.00   
2               2.00               1.00            0.00            0.00   
3               2.00               1.00            0.00            0.00   
4               2.00               2.00            0.00            0.00   

   home_fouls_conceded  away_fouls_conceded  home_SoT_%  away_SoT_%  \
0                11.00                 9.00       33.33       10.00   
1                 9.00                11.00       40.00       83.33   
2                 9.00                 8.00       16.67       30.77   
3                11.00                 8.00       75.00       30.77   
4                11.00                12.00       13.33       33.33   

   home_QuanCR  away_QuanCR  home_QualCR  away_QualCR  home_ppg  away_ppg  \
0        16.67         0.00        50.00         0.00      3.00      0.00   
1         0.00        16.67         0.00        20.00      0.00      3.00   
2         0.00        23.08         0.00        75.00      0.00      3.00   
3        25.00         7.69        33.33        25.00      3.00      0.00   
4         6.67        13.33        50.00        40.00      0.00      3.00   

   home_win_rate  away_win_rate  
0         100.00           0.00  
1           0.00         100.00  
2           0.00         100.00  
3         100.00           0.00  
4           0.00         100.00

### Pre Lockdown and Post Lockdown Data Manipulation

Last matchweek before covid-19 attendance/crowd prohibition: 
   Matchweek 29: 10 March 2020 Leicester vs Aston Villa

In [9]:
#last match before covid-19 suspension
last_idx = df.loc[df['date']=='2020-03-10'].index[0]

In [10]:
#partitioning data before and after covid-19 suspension
pre_covid_data = df[:last_idx+1]
post_covid_data = df[last_idx+1:]

To make analysis from 3 EPL seasons, we only take teams that played in those all 3 seasons and eliminate data of teams that didn't.

2018/2019 relegated teams:
- Fulham
- Cardiff City
- Huddersfield Town

2019/2020 promoted teams:
- Norwich City
- Sheffield Utd
- Aston Villa

2019/2020 relegated teams:
- Bournemouth
- Watford
- Norwich City

2020/2021 promoted teams:
- Leeds Utd
- West Brom
- Fulham

In [11]:
all_teams = list(df['home_team'].unique())
relegated_promoted_teams = ['Fulham', 'Huddersfield Town', 'Cardiff City', 
                  'Norwich City', 'Sheffield United', 'Aston Villa', 
                  'AFC Bournemouth', 'Watford', 'Leeds United', 'West Bromwich Albion']

#stayed_teams = list(set(all_teams)-set(relegated_promoted_teams))
stayed_teams = [team for team in all_teams if team not in relegated_promoted_teams]

In [12]:
'''home_pre_covid_pts = pre_covid_data.groupby(['home_team']).agg(avg_pts_before = ('home_ppg','mean')).loc[stayed_teams].reset_index()
home_post_covid_pts = post_covid_data.groupby(['home_team']).agg(avg_pts_after = ('home_ppg','mean')).loc[stayed_teams].reset_index()
home_df = home_pre_covid_pts.merge(home_post_covid_pts,on='home_team')
home_df'''

"home_pre_covid_pts = pre_covid_data.groupby(['home_team']).agg(avg_pts_before = ('home_ppg','mean')).loc[stayed_teams].reset_index()\nhome_post_covid_pts = post_covid_data.groupby(['home_team']).agg(avg_pts_after = ('home_ppg','mean')).loc[stayed_teams].reset_index()\nhome_df = home_pre_covid_pts.merge(home_post_covid_pts,on='home_team')\nhome_df"

In [13]:
'''away_pre_covid_pts = pre_covid_data.groupby(['away_team']).agg(avg_pts_before = ('away_ppg','mean')).loc[stayed_teams].reset_index()
away_post_covid_pts = post_covid_data.groupby(['away_team']).agg(avg_pts_after = ('away_ppg','mean')).loc[stayed_teams].reset_index()
away_df = away_pre_covid_pts.merge(away_post_covid_pts,on='away_team')
away_df'''

"away_pre_covid_pts = pre_covid_data.groupby(['away_team']).agg(avg_pts_before = ('away_ppg','mean')).loc[stayed_teams].reset_index()\naway_post_covid_pts = post_covid_data.groupby(['away_team']).agg(avg_pts_after = ('away_ppg','mean')).loc[stayed_teams].reset_index()\naway_df = away_pre_covid_pts.merge(away_post_covid_pts,on='away_team')\naway_df"

In [44]:
def home_stats(stat, team, summary):
    #aggregating team playing at home before and after lockdown
    home_pre_covid_stat = pre_covid_data.groupby(['home_team']).agg({stat:'mean'}).loc[team].reset_index()
    home_pre_covid_stat = home_pre_covid_stat.rename(columns={stat:'{}_before'.format(stat)})
    home_post_covid_stat = post_covid_data.groupby(['home_team']).agg({stat:'mean'}).loc[team].reset_index()
    home_post_covid_stat = home_post_covid_stat.rename(columns={stat:'{}_after'.format(stat)})
    
    #merge df of before and after lockdown
    home_df = home_pre_covid_stat.merge(home_post_covid_stat,on='home_team')
    
    #extract the 'changes' of team performance before and after lockdown when playing at home
    home_df['changes'] = home_df.iloc[:,2] - home_df.iloc[:,1]
    
    #sort dataframe by 'changes' ascending
    home_df = home_df.sort_values(by='changes', ascending=True).reset_index(drop=True)
    
    #Summary of average metric changes of all teams
    if summary==True:
        avg_change = round(sum(home_df['changes'])/len(home_df['changes']),2)
        print('Average {} changes of home team is {}'.format(stat, avg_change))
    
    return home_df

In [30]:
def away_stats(stat,team):
    #aggregating team playing at away match before and after lockdown
    away_pre_covid_stat = pre_covid_data.groupby(['away_team']).agg({stat:'mean'}).loc[team].reset_index()
    away_pre_covid_stat = away_pre_covid_stat.rename(columns={stat:'{}_before'.format(stat)})
    away_post_covid_stat = post_covid_data.groupby(['away_team']).agg({stat:'mean'}).loc[team].reset_index()
    away_post_covid_stat = away_post_covid_stat.rename(columns={stat:'{}_after'.format(stat)})
    
    #merge df of before and after lockdown
    away_df = away_pre_covid_stat.merge(away_post_covid_stat,on='away_team')
    
    #extract the 'changes' of team performance before and after lockdown when playing at away stadium
    away_df['changes'] = away_df.iloc[:,2] - away_df.iloc[:,1]
    #sort dataframe by 'changes' ascending
    away_df = away_df.sort_values(by='changes', ascending=True).reset_index(drop=True)
    
    #Summary of average metric changes of all teams
    if summary==True:
        avg_change = round(sum(away_df['changes'])/len(away_df['changes']),2)
        print('Average {} of away team is {}'.format(stat, avg_change))
    
    return away_df

#### Points per Game

In [47]:
home_ppg = home_stats('home_ppg', stayed_teams, summary=True)
home_ppg

Average home_ppg changes of home team is -0.21


home_team  home_ppg_before  home_ppg_after  changes
0                  Liverpool             2.88            1.87    -1.01
1                    Everton             1.79            1.17    -0.62
2                    Arsenal             2.09            1.65    -0.44
3                    Burnley             1.35            1.00    -0.35
4            Manchester City             2.59            2.36    -0.23
5           Newcastle United             1.39            1.17    -0.23
6          Manchester United             1.91            1.70    -0.22
7   Brighton and Hove Albion             1.24            1.04    -0.20
8             Crystal Palace             1.24            1.04    -0.19
9    Wolverhampton Wanderers             1.65            1.48    -0.17
10            Leicester City             1.68            1.52    -0.15
11         Tottenham Hotspur             1.94            1.92    -0.02
12                   Chelsea             1.94            1.96     0.02
13           West Ham United             1.39            1.71     0.31
14               Southampton             1.09            1.48     0.39

Information obtained from home_ppg dataframe:
- 12 out of 15 (80%) clubs had decreased points per game, averaging at -0.21 points per game after crowd suspension when playing at home ground
- Liverpool suffers the most from crowd suspension, having decreased 1.01 points per game
- Tottenham Hotspur and Chelsea home points per game did not get affected by crowd suspension with only slightly changes 

In [31]:
away_ppg = away_stats('away_ppg', stayed_teams, summary=True)
away_ppg

Average away_ppg of away team is 0.19


away_team  away_ppg_before  away_ppg_after  changes
0                  Liverpool             2.45            1.79    -0.66
1             Crystal Palace             1.39            1.00    -0.39
2    Wolverhampton Wanderers             1.33            1.12    -0.21
3                    Chelsea             1.64            1.67     0.03
4                Southampton             1.09            1.12     0.03
5          Tottenham Hotspur             1.41            1.48     0.07
6            Manchester City             2.12            2.22     0.10
7           Newcastle United             1.00            1.13     0.13
8             Leicester City             1.45            1.67     0.21
9                    Burnley             1.00            1.29     0.29
10                   Arsenal             1.22            1.56     0.34
11  Brighton and Hove Albion             0.71            1.22     0.51
12           West Ham United             0.97            1.57     0.59
13                   Everton             0.94            1.87     0.93
14         Manchester United             1.39            2.33     0.94

Information obtained from away_ppg dataframe:
- The crowd suspension effect is parallel.  As a result of dropping points per game for home team, 12 out of 15 clubs had increased points per game, averaging at +0.19 points per game after crowd suspension when playing at away ground
- Unlike majority of clubs that got this increasing away ppg advantage, 3 clubs; Liverpool, Wolverhampton Wanderers and Crystal Palace  were having decreased away ppg with Liverpool had the most decreased ppg (-0.66)
- Manchester United gained the most points per game (+0.94) from this away ppg advantage.

#### Shots on Target

In [18]:
home_SoT = home_stats('home_SoT_%', stayed_teams, summary=True)
home_SoT 

Average home_SoT_% changes of home team is 2.1


home_team  home_SoT_%_before  home_SoT_%_after  changes
0          Manchester United              44.37             37.55    -6.82
1                    Burnley              34.64             31.61    -3.03
2                  Liverpool              39.05             37.29    -1.76
3            West Ham United              39.46             37.92    -1.54
4                    Everton              34.06             33.68    -0.38
5   Brighton and Hove Albion              27.46             28.17     0.71
6             Leicester City              35.16             36.32     1.16
7                    Arsenal              37.01             39.07     2.06
8    Wolverhampton Wanderers              31.02             33.19     2.17
9            Manchester City              35.74             38.94     3.20
10               Southampton              36.56             40.81     4.25
11                   Chelsea              36.39             41.24     4.84
12          Newcastle United              30.58             36.46     5.88
13         Tottenham Hotspur              35.27             42.51     7.25
14            Crystal Palace              27.36             40.80    13.45

Information obtained from home_SoT (shots on target percentage of home team) dataframe:
- Average changes of shots on target percentage of home team is 2.1 % 
- 2 teams who had the most significant changes of SoT percentage at home are Crystal Palace and Manchester United. Crystal Palace SoT percentage significantly increased and Manchester United SoT percentage significantly dropped.

In [19]:
away_SoT= away_stats('away_SoT_%', stayed_teams, summary=True)
away_SoT

Average away_SoT_% of away team is 0.03


away_team  away_SoT_%_before  away_SoT_%_after  changes
0          Tottenham Hotspur              40.65             34.87    -5.78
1            West Ham United              37.73             33.06    -4.68
2                  Liverpool              39.90             36.62    -3.29
3            Manchester City              40.14             37.85    -2.30
4                    Arsenal              36.90             34.64    -2.26
5             Crystal Palace              37.69             35.48    -2.22
6             Leicester City              39.59             38.06    -1.53
7   Brighton and Hove Albion              33.80             33.08    -0.72
8    Wolverhampton Wanderers              32.64             32.74     0.10
9                Southampton              37.03             40.20     3.17
10         Manchester United              39.96             43.61     3.65
11          Newcastle United              33.48             37.22     3.74
12                   Chelsea              32.08             35.87     3.78
13                   Everton              35.13             39.46     4.34
14                   Burnley              30.85             35.28     4.43

Information obtained from home_SoT (shots on target percentage of home team) dataframe:
- Average changes of shots on target percentage of away team is 0.03 %. 
- 2 teams who had the most significant changes of SoT percentage at away are Burnley and Tottenham Hotspur. Burnley SoT percentage significantly increased and Tottenham Hotspur SoT percentage significantly dropped

#### Quantity Conversion Rate

Quantity conversion rate is goal conversion rate from all shots created (shots on target + shots off target).

In [20]:
home_QuanCR = home_stats('home_QuanCR', stayed_teams, summary=True)
home_QuanCR

Average home_QuanCR changes of home team is 1.16


home_team  home_QuanCR_before  home_QuanCR_after  changes
0                  Liverpool               16.65              12.62    -4.03
1    Wolverhampton Wanderers               12.47               8.64    -3.84
2                    Burnley               11.96               8.14    -3.82
3   Brighton and Hove Albion               10.58               8.50    -2.08
4                    Arsenal               15.34              14.33    -1.01
5                    Chelsea               10.69              11.13     0.44
6          Manchester United               13.12              13.98     0.86
7             Leicester City               11.52              12.54     1.01
8                    Everton               10.22              11.94     1.71
9            Manchester City               14.21              16.17     1.95
10               Southampton               10.54              12.73     2.19
11         Tottenham Hotspur               14.61              17.66     3.04
12          Newcastle United                9.47              13.46     3.99
13           West Ham United               12.68              17.62     4.94
14            Crystal Palace                6.59              18.67    12.09

Information obtained from home_QuanCR (quantity conversion rate of home team) dataframe:

- Average changes of quantity conversion rate of home team is 1.16%.
- Crystal Palace has the most significantly increased quantity conversion rate at home after crowd suspension (12.09%)
- Liverpool has the most significantly dropped quantity conversion rate at home after crowd suspension (-4.03%)

In [21]:
away_QuanCR = away_stats('away_QuanCR', stayed_teams, summary=True)
away_QuanCR

Average away_QuanCR of away team is -0.47


away_team  away_QuanCR_before  away_QuanCR_after  changes
0                    Burnley               16.21               9.09    -7.11
1             Crystal Palace               17.15              11.49    -5.66
2    Wolverhampton Wanderers               11.97               7.75    -4.22
3                    Arsenal               16.29              14.22    -2.08
4                  Liverpool               14.07              12.55    -1.51
5                    Chelsea               10.90              10.99     0.09
6           Newcastle United               10.09              10.36     0.27
7          Tottenham Hotspur               14.91              15.44     0.53
8   Brighton and Hove Albion                9.82              10.42     0.60
9             Leicester City               13.87              14.79     0.92
10               Southampton               10.14              11.13     0.99
11           Manchester City               13.88              15.65     1.77
12           West Ham United                9.65              11.47     1.82
13                   Everton               10.49              13.05     2.55
14         Manchester United               11.76              15.73     3.97

Information obtained from away_QuanCR (quantity conversion rate of away team) dataframe:
- Average changes of quantity conversion rate of home team is -0.47%.
- Manchester United has the most significantly increased quantity conversion rate at away after crowd suspension (3.97%)
- Burnley has the most significantly dropped quantity conversion rate at home after crowd suspension (-7.11%)

#### Quality Conversion Rate

Quality conversion rate is goal conversion rate from shots on target (excluding shots off target).

In [22]:
home_QualCR = home_stats('home_QualCR', stayed_teams, summary=True)
home_QualCR 

Average home_QualCR changes of home team is -0.86


home_team  home_QualCR_before  home_QualCR_after  changes
0                    Burnley               39.20              23.34   -15.85
1                  Liverpool               43.14              29.11   -14.03
2    Wolverhampton Wanderers               38.27              27.72   -10.55
3                    Arsenal               42.15              32.54    -9.62
4   Brighton and Hove Albion               32.88              26.25    -6.63
5                    Chelsea               32.38              27.25    -5.13
6                Southampton               30.21              30.14    -0.07
7             Leicester City               31.82              32.37     0.55
8           Newcastle United               29.05              30.61     1.56
9            Manchester City               38.18              40.66     2.48
10         Tottenham Hotspur               37.67              41.83     4.16
11         Manchester United               27.96              35.87     7.92
12                   Everton               31.84              40.60     8.76
13           West Ham United               31.87              41.57     9.69
14            Crystal Palace               23.80              37.64    13.85

Information obtained from home_QualCR (quality conversion rate of home team) dataframe:
- Average changes of quality conversion rate of home team is -0.86%.
- Crystal Palace has the most significantly increased quality conversion rate at home after crowd suspension (13.85%)
- Burnley and Liverpool have the most significantly dropped quality conversion rate at home after crowd suspension (-15.85% and -14.03% respectively)

In [23]:
away_QualCR = away_stats('away_QualCR', stayed_teams, summary=True)
away_QualCR

Average away_QualCR of away team is -0.74


away_team  away_QualCR_before  away_QualCR_after  changes
0                    Burnley               37.83              23.28   -14.54
1             Crystal Palace               42.34              28.53   -13.81
2    Wolverhampton Wanderers               32.07              23.06    -9.02
3                    Arsenal               40.90              34.65    -6.26
4                    Chelsea               35.21              30.73    -4.48
5                  Liverpool               34.80              32.88    -1.92
6           Newcastle United               30.13              29.38    -0.75
7             Leicester City               34.25              35.37     1.12
8                    Everton               28.04              29.83     1.80
9                Southampton               25.36              28.67     3.31
10         Manchester United               29.93              34.70     4.77
11         Tottenham Hotspur               36.28              41.16     4.88
12  Brighton and Hove Albion               22.79              30.07     7.29
13           Manchester City               32.87              40.71     7.84
14           West Ham United               21.63              30.23     8.60

Information obtained from home_QualCR (quality conversion rate of home team) dataframe:
- Average changes of quality conversion rate of away team is -0.74%.
- West Ham has the most significantly increased quality conversion rate at away after crowd suspension (8.60%)
- Burnley and Crystal Palace have the most significantly dropped quality conversion rate at away after crowd suspension (-14.54% and -13.81% respectively)

In [24]:
home_stats('home_win_rate', stayed_teams, summary=True)

Average home_win_rate changes of home team is -7.31


home_team  home_win_rate_before  home_win_rate_after  \
0                  Liverpool                 94.12                56.52   
1                    Everton                 51.52                29.17   
2                    Burnley                 41.18                21.74   
3                    Arsenal                 61.76                47.83   
4           Newcastle United                 39.39                29.17   
5   Brighton and Hove Albion                 30.30                20.83   
6            Manchester City                 84.38                76.00   
7             Crystal Palace                 32.35                26.09   
8          Manchester United                 52.94                47.83   
9          Tottenham Hotspur                 60.61                58.33   
10            Leicester City                 50.00                47.83   
11   Wolverhampton Wanderers                 44.12                43.48   
12                   Chelsea                 55.88                56.52   
13           West Ham United                 39.39                50.00   
14               Southampton                 26.47                43.48   

    changes  
0    -37.60  
1    -22.35  
2    -19.44  
3    -13.94  
4    -10.23  
5     -9.47  
6     -8.38  
7     -6.27  
8     -5.12  
9     -2.27  
10    -2.17  
11    -0.64  
12     0.64  
13    10.61  
14    17.01

In [25]:
away_stats('away_win_rate', stayed_teams)

Average away_win_rate of away team is 6.44


away_team  away_win_rate_before  away_win_rate_after  \
0                  Liverpool                 75.76                50.00   
1             Crystal Palace                 39.39                29.17   
2    Wolverhampton Wanderers                 33.33                29.17   
3          Tottenham Hotspur                 41.18                39.13   
4                Southampton                 30.30                29.17   
5                    Chelsea                 48.48                50.00   
6            Manchester City                 67.65                69.57   
7             Leicester City                 42.42                45.83   
8           Newcastle United                 23.53                30.43   
9                    Burnley                 24.24                37.50   
10  Brighton and Hove Albion                 14.71                30.43   
11           West Ham United                 26.47                43.48   
12                   Arsenal                 28.12                48.00   
13         Manchester United                 39.39                66.67   
14                   Everton                 23.53                56.52   

    changes  
0    -25.76  
1    -10.23  
2     -4.17  
3     -2.05  
4     -1.14  
5      1.52  
6      1.92  
7      3.41  
8      6.91  
9     13.26  
10    15.73  
11    17.01  
12    19.88  
13    27.27  
14    32.99

#### Possession

In [26]:
home_poss = home_stats('home_possession_%', stayed_teams, summary=True)
home_poss

Average home_possession_% changes of home team is -0.68


home_team  home_possession_%_before  \
0          Tottenham Hotspur                     58.73   
1                    Arsenal                     59.00   
2                    Everton                     52.55   
3             Crystal Palace                     45.98   
4            West Ham United                     47.63   
5                    Chelsea                     62.71   
6            Manchester City                     67.76   
7             Leicester City                     55.70   
8           Newcastle United                     41.60   
9   Brighton and Hove Albion                     49.81   
10                 Liverpool                     62.85   
11                   Burnley                     42.55   
12         Manchester United                     55.76   
13   Wolverhampton Wanderers                     48.48   
14               Southampton                     45.49   

    home_possession_%_after  changes  
0                     50.82    -7.91  
1                     52.74    -6.26  
2                     47.93    -4.63  
3                     41.80    -4.18  
4                     43.60    -4.03  
5                     60.49    -2.21  
6                     66.72    -1.04  
7                     55.65    -0.04  
8                     41.93     0.33  
9                     51.43     1.61  
10                    64.78     1.93  
11                    44.67     2.12  
12                    58.85     3.09  
13                    52.39     3.91  
14                    52.62     7.13

In [35]:
away_stats('away_possession_%', stayed_teams, summary=True)

Average away_possession_% of away team is 0.05


away_team  away_possession_%_before  \
0                    Everton                     49.26   
1            Manchester City                     66.74   
2            West Ham United                     45.55   
3             Crystal Palace                     43.43   
4                    Arsenal                     55.41   
5          Tottenham Hotspur                     54.25   
6                  Liverpool                     62.07   
7   Brighton and Hove Albion                     46.15   
8           Newcastle United                     35.99   
9          Manchester United                     53.54   
10                   Chelsea                     61.21   
11                   Burnley                     38.42   
12   Wolverhampton Wanderers                     45.12   
13            Leicester City                     51.19   
14               Southampton                     44.78   

    away_possession_%_after  changes  
0                     45.14    -4.12  
1                     62.74    -4.00  
2                     42.45    -3.10  
3                     40.45    -2.98  
4                     52.71    -2.69  
5                     51.94    -2.31  
6                     61.16    -0.91  
7                     46.90     0.75  
8                     36.77     0.78  
9                     54.75     1.22  
10                    62.85     1.64  
11                    40.45     2.03  
12                    47.54     2.43  
13                    55.57     4.37  
14                    52.43     7.65

## Visualization

In [51]:
def home_stat_viz(stat, team, summary):
    fig = go.Figure(data=[
        go.Bar(name='ppg before covid 19', 
               x=home_stats(stat,team, summary)['home_team'], 
               y=home_stats(stat,team, summary)['home_ppg_before'],
               text=home_stats(stat,team, summary)['home_ppg_before'], 
               textposition='auto',
               texttemplate='%{text:.3s}'),
         go.Bar(name='ppg after covid 19', 
                x=home_stats(stat,team, summary)['home_team'], 
                y=home_stats(stat,team, summary)['home_ppg_after'],
                text=home_stats(stat,team, summary)['home_ppg_after'],
                textposition='auto',
                texttemplate='%{text:.3s}')
    ])
    # Change the bar mode
    fig.update_layout(barmode='group', title_text='Points per Game of Home Game')
    fig.show()

In [52]:
home_stat_viz('home_ppg', stayed_teams, summary=False)